install/kubernetes/cilium/templates/cilium-operator/deployment.yaml

```diff
@@ -18,10 +18,22 @@ spec:
     matchLabels:
       io.cilium/app: operator
       name: cilium-operator
+  # ensure operator update on single node k8s clusters, by using rolling update with maxUnavailable=100% in case
+  # of one replica and no user configured Recreate strategy.
+  # otherwise an update might get stuck due to the default maxUnavailable=50% in combination with the
+  # podAntiAffinity which prevents deployments of multiple operator replicas on the same node.
+  {{- if and (eq (.Values.operator.replicas | toString) "1") (eq .Values.operator.updateStrategy.type "RollingUpdate") }}
+  strategy:
+    rollingUpdate:
+      maxSurge: {{ .Values.operator.updateStrategy.rollingUpdate.maxSurge }}
+      maxUnavailable: 100%
```

Comment:
Why is there helm value vars for this when we are hardcoding it to 100%?

Reply:
we're only doing this if there's only 1 replica of the operator and the user hasn't decided to switch to a Recreate update strategy. we assume that in these cases, the user isn't interested in any HA and therefore proper rolling update feature.

As mentioned, this is necessary to ensure, that it's still possible to update the operator on single node k8s clusters.

Without this customization, the operator update might get stuck on single node clusters due to the default maxUnavailable=50% in combination with the podAntiAffinity which prevents multiple operator replicas to be deployed on the same k8s node. the Pod from the old deployment wouldn't be terminated and the one from the new deployment can't be scheduled...

Reply:
Oh I see, this is just for non HA setups. Makes sense.

Reply:
One thing about this is if the user decides to change their replica count using `kubectl scale ... --replicas=2` later on then the incorrect rollout strategy will remain.

Reply:
Not sure if thats really a case we need to cover

Reply:
it's definitely a good point which might confuses users. but at least the default behaviour of the helm chart (which uses 2 replicas) should be covered correctly.